tests/test_modeling_reformer.py

```diff
@@ -1049,8 +1120,23 @@ def test_pretrained_generate_crime_and_punish(self):
         output_ids = model.generate(
             input_ids, max_length=50, num_beams=4, early_stopping=True, do_sample=False, num_hashes=8
         )
-        output_text = tokenizer.decode(output_ids[0])
+        output = tokenizer.decode(output_ids[0])
+
         self.assertEqual(
-            output_text,
+            output,
             "A few months later state expression in his ideas, at the first entrance. He was positively for an inst",
         )
+
+    @slow
+    def test_pretrained_generate_use_cache_equality(self):
+        model = ReformerModelWithLMHead.from_pretrained("google/reformer-crime-and-punishment").to(torch_device)
+        tokenizer = ReformerTokenizer.from_pretrained("google/reformer-crime-and-punishment")
+        model.eval()
+        input_ids = tokenizer.encode("A few months later", return_tensors="pt").to(torch_device)
+        output_ids_with_cache = model.generate(input_ids, max_length=130, num_hashes=8, use_cache=False)
```

Comment:
need `num_hashes=8` to ensure equal output.

Reply:
why?

consider testing some bad inputs and making sure you raise the errors you want.

This is a lot of code for one test, albeit a strong one. Do you need a big model to run this test though?

Reply:
Yeah, good question. So in general using the `cache` vs. not using the cache is never 100% the same because when one does not use the cache, the values are masked by -10^5 which is not the same as not calculating them at all. That's why all our `use_cache == no_use_cache` tests compare on 10e-3 level because at 10e-5 they would already fail (tested this with T5). 

In the case of Reformer, it is even trickier because there is a `self_mask` (masked the token of the same position) and a `causal_mask` which are different (1e-5 for `self_mask` and 1e-9 for `causal_mask`) because there are cases where all tokens have 1e-9 and only the first position has 1e-5 (this happens for the first token of the sequence for example) => therefore the values are much more sensible to removing the masks completely when doing cached inference.

`num_hashes` remedies this because it averages over `num_hashes` query-key calculations so the results become more stable. No need to run this test on a big model though.